From 4bac1b81fb3b5a512d004492ec6bc045c30224a6 Mon Sep 17 00:00:00 2001
From: robcore <robpatershuk@gmail.com>
Date: Thu, 28 Apr 2016 16:02:59 -0600
Subject: [PATCH 25/50] From ddeec2f56c03f9b6c017478f288faab8eb6eb678 Mon Sep
 17 00:00:00 2001 From: flar2 <asegaert@gmail.com> Date: Fri, 20 Sep 2013
 23:00:13 -0400 Subject: [PATCH 038/292] freezer: add new freezable helpers
 using  freezer_do_not_count

Signed-off-by: flar2 <asegaert@gmail.com>
---
 include/linux/freezer.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/include/linux/freezer.h b/include/linux/freezer.h
index d9608f1..c1b49cb 100644
--- a/include/linux/freezer.h
+++ b/include/linux/freezer.h
@@ -161,6 +161,15 @@ static inline int freezer_should_skip(struct task_struct *p)
 	__retval;							\
 })

+#define wait_event_freezable_exclusive(wq, condition)		\
+({									\
+	int __retval;							\
+	freezer_do_not_count();						\
+	__retval = wait_event_interruptible_exclusive(wq, condition);	\
+	freezer_count();						\
+	__retval;							\
+})
+
 #else /* !CONFIG_FREEZER */
 static inline bool frozen(struct task_struct *p) { return false; }
 static inline bool freezing(struct task_struct *p) { return false; }
--
2.8.0.rc3
---
 include/linux/freezer.h      |  9 +++++++++
 include/linux/freezer.h.orig | 22 +++++++---------------
 2 files changed, 16 insertions(+), 15 deletions(-)

diff --git a/include/linux/freezer.h b/include/linux/freezer.h
index 577bc02..27a397e 100644
--- a/include/linux/freezer.h
+++ b/include/linux/freezer.h
@@ -206,6 +206,15 @@ static inline bool freezer_should_skip(struct task_struct *p)
 	__retval;							\
 })
 
+#define wait_event_freezable_exclusive(wq, condition)		\
+({									\
+	int __retval;							\
+	freezer_do_not_count();						\
+	__retval = wait_event_interruptible_exclusive(wq, condition);	\
+	freezer_count();						\
+	__retval;							\
+})
+
 #else /* !CONFIG_FREEZER */
 static inline bool frozen(struct task_struct *p) { return false; }
 static inline bool freezing(struct task_struct *p) { return false; }
diff --git a/include/linux/freezer.h.orig b/include/linux/freezer.h.orig
index 364fd51..577bc02 100644
--- a/include/linux/freezer.h.orig
+++ b/include/linux/freezer.h.orig
@@ -190,27 +190,19 @@ static inline bool freezer_should_skip(struct task_struct *p)
 #define wait_event_freezable(wq, condition)				\
 ({									\
 	int __retval;							\
-	for (;;) {							\
-		__retval = wait_event_interruptible(wq, 		\
-				(condition) || freezing(current));	\
-		if (__retval || (condition))				\
-			break;						\
-		try_to_freeze();					\
-	}								\
+	freezer_do_not_count();						\
+	__retval = wait_event_interruptible(wq, (condition));		\
+	freezer_count();						\
 	__retval;							\
 })
 
 #define wait_event_freezable_timeout(wq, condition, timeout)		\
 ({									\
 	long __retval = timeout;					\
-	for (;;) {							\
-		__retval = wait_event_interruptible_timeout(wq,		\
-				(condition) || freezing(current),	\
-				__retval); 				\
-		if (__retval <= 0 || (condition))			\
-			break;						\
-		try_to_freeze();					\
-	}								\
+	freezer_do_not_count();						\
+	__retval = wait_event_interruptible_timeout(wq,	(condition),	\
+				__retval);				\
+	freezer_count();						\
 	__retval;							\
 })
 
-- 
2.8.0.rc3

